feat: rewind keyword search

shaik-zeeshan opened the pull request:
related: #1271 
/claim #1271



https://github.com/user-attachments/assets/57bfe64f-5dd3-47c6-bb0f-44b4c1eb21d7


https://github.com/user-attachments/assets/952f0e89-101e-4957-897d-780efc753446

louis030195 commented on screenpipe-server/src/db.rs:
```diff
@@ -2133,4 +2135,172 @@ impl DatabaseManager {
+
+        // Add conditions for each keyword
+        if fuzzy_match {
+            conditions.extend(keywords.iter().map(|_| "o.text LIKE '%' || ? || '%'"));
```
isn't the FTS 100000x faster than LIKE?

`O(log(n))`

shaik-zeeshan replied:
yeah you are right 

changing the implementation to use FTS

louis030195 commented:
trying to understand 

is it full refactor of timeline or different page in it?

like ideas here, however the query is way too slow to generate suggestion, most people don't have this attention span 

how do you extract these logos? this is quite interesting as some people ask for "universal file search" or like typical use case:

"ok im trying to find my rent document is it in dropbox, google drive, gmail, icloud, somewhere else...?"

and also this is a use case for teams, for example if my friend saw a document and i want to find it, it would be so useful to be able to search and find it in seconds 

#1293 related

this is kinda out of scope of the initial search idea i wrote but kind of extension of it 

we've also explored idea of using models like https://github.com/microsoft/OmniParser to index images/logos more precisely than OCR for the future FYI


PS: cannot test right now - focused on #1318

shaik-zeeshan commented:
> is it full refactor of timeline or different page in it?

it is different page

> however the query is way too slow to generate suggestion, most people don't have this attention span

using local model for suggestions, maybe I could generate suggestions based on app_name without using models.
so it doesn't too long

> how do you extract these logos? this is quite interesting as some people ask for "universal file search" or like typical use case:

getting links using text and fetching favicon from google 

> "ok im trying to find my rent document is it in dropbox, google drive, gmail, icloud, somewhere else...?"
> 
> and also this is a use case for teams, for example if my friend saw a document and i want to find it, it would be so useful to be able to search and find it in seconds

this is cool, maybe we could use models to get related info or use embeddings for this

louis030195 commented:
hows the AI settings config?

<img width="1512" alt="Screenshot 2025-02-13 at 8 38 39 AM" src="https://github.com/user-attachments/assets/37eba80b-27a4-4c03-bbdf-52cbc93256e2" />
<img width="1512" alt="Screenshot 2025-02-13 at 8 38 23 AM" src="https://github.com/user-attachments/assets/b7a77333-9806-4d63-a74d-d33074d42702" />

louis030195 commented:
https://github.com/user-attachments/assets/c6066c1f-4def-4cc2-b6d2-c291cb2be39b


also would appreciate if u can fix the top bar navigation

shaik-zeeshan commented:
> <img alt="Screenshot 2025-02-13 at 8 38 23 AM" width="1512" src="https://private-user-images.githubusercontent.com/25003283/412981728-b7a77333-9806-4d63-a74d-d33074d42702.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk0NjUzNTYsIm5iZiI6MTczOTQ2NTA1NiwicGF0aCI6Ii8yNTAwMzI4My80MTI5ODE3MjgtYjdhNzczMzMtOTgwNi00ZDYzLWE3NGQtZDMzMDc0ZDQyNzAyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEzVDE2NDQxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFmZWUyZjcwODc3OTJlZTUxYjZkZjBmNDU1NGRmNTAwYWViN2IzMmFhYTI1NzJiNWEwZWQyNjY0MzViMmY1NjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tjk3mvMM5JjTFocBQ3YZeDgCL7FpnrEOXbEfSMx1uL8">

looking into this


> also would appreciate if u can fix the top bar navigation

sure

louis030195 commented:
just tested finally

1. can we scroll the same order of the timeline? eg things are from right to left, scrolling down goes left (in the past)

<img width="1312" alt="Screenshot 2025-02-15 at 11 51 03 AM" src="https://github.com/user-attachments/assets/44bf7dfa-0b16-44e5-9976-a2aafc9c1020" />

2. is there a way to highlight on the image in yellow the text, like rewind does? and make it selectable? (how hard would it be? could be another PR)

3. sometimes frames showed up even though word "matt" was not on the frame, is that a backend issue or?

4. did not really understand the query generation, also is there a way to see which model is used like robot icon i did in other ui?


5. maybe add icon to go back to timeline? back arrow or something else
6. did not understand clearly the date input at first on search page
7. did not see this feature after a while
<img width="200" alt="image" src="https://github.com/user-attachments/assets/46b8f7d2-ec2e-490f-9833-cc783cfd4840" />

8. am a bit concerned by this:

<img width="1512" alt="Screenshot 2025-02-15 at 11 57 41 AM" src="https://github.com/user-attachments/assets/5a4672d2-bca5-417e-a49e-3fcc100db206" />

do we have an issue in backend where the frame is desync with the content/ocr/etc.?

i think this feature is not great rn, not sure how to make it better but either hide it for now or figure out something that make sense, how often would people want to get description of the image? maybe can be a way to extract text better than OCR using multimodal? or idk



other issues i encountered:


<img width="1512" alt="Screenshot 2025-02-15 at 11 55 50 AM" src="https://github.com/user-attachments/assets/d26921c2-054b-4d89-8b21-482aea245c81" />
<img width="1512" alt="Screenshot 2025-02-15 at 11 55 38 AM" src="https://github.com/user-attachments/assets/355db44e-e7d3-4391-b99b-c1f73838b3fa" />
<img width="1512" alt="Screenshot 2025-02-15 at 11 55 31 AM" src="https://github.com/user-attachments/assets/bbb4fe1a-2c20-4689-9991-ba35cfa81d88" />
<img width="1512" alt="Screenshot 2025-02-15 at 11 55 24 AM" src="https://github.com/user-attachments/assets/d6f50b94-655c-4efd-8901-8a77247dea19" />
<img width="1512" alt="Screenshot 2025-02-15 at 11 55 08 AM" src="https://github.com/user-attachments/assets/b79ffe40-f84a-46c8-a00a-20bba2d57704" />
<img width="1512" alt="Screenshot 2025-02-15 at 11 54 55 AM" src="https://github.com/user-attachments/assets/68ab6eac-b3a6-407f-a8cd-32531afa2df0" />

i think the suggestion query is kinda awkward for now, maybe we could tune down the scope of this PR to just simple full text search over OCR/transcriptions or iterate on this UX further as you prefer 


overall this is really great PR already, just need some adjustment or tuning down the scope as you prefer, happy to increase the bounty or add tip!

shaik-zeeshan commented:
> 1. can we scroll the same order of the timeline? eg things are from right to left, scrolling down goes left (in the past)

sure

> 2. is there a way to highlight on the image in yellow the text, like rewind does? and make it selectable? (how hard would it be? could be another PR)

I already did this in code but comment it out because it wasn't accurate 

https://github.com/user-attachments/assets/9902b52f-74d3-49f9-a36b-12300a611224


> 3\. sometimes frames showed up even though word "matt" was not on the frame, is that a backend issue or?

yeah sometimes frame doesn't match content and app_name and window_name


> 5\. maybe add icon to go back to timeline? back arrow or something else

I will add this

> 6\. did not understand clearly the date input at first on search page

will make it simpler and make placeholder be something understandable


> 8\. am a bit concerned by this:
> 
> <img alt="Screenshot 2025-02-15 at 11 57 41 AM" width="1512" src="https://private-user-images.githubusercontent.com/25003283/413570230-5a4672d2-bca5-417e-a49e-3fcc100db206.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk2ODA5NjMsIm5iZiI6MTczOTY4MDY2MywicGF0aCI6Ii8yNTAwMzI4My80MTM1NzAyMzAtNWE0NjcyZDItYmNhNS00MTdlLWE0OWUtM2ZjYzEwMGRiMjA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE2VDA0Mzc0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0NmYzOGJmNmM4YTNlNjU4NGY5YjkxMDcyM2M0NTBlYWE5YmU3ZGRlYzdiMzVkMjlkN2QwOGEwNGNhMDA2OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.wO2J0Nqn33LZG9g9xJfTgx4QjZl_P4WdT9oX4In8bPo">
> i think this feature is not great rn, not sure how to make it better but either hide it for now or figure out something that make sense, how often would people want to get description of the image? maybe can be a way to extract text better than OCR using multimodal? or idk

we could remove this since the user can see the frame. so it doesn't really feel useful . I only did this because I wanted to show links to pages in frame. currently we don't store any links. showing links to the page present in frame will be a useful feature


> <img alt="Screenshot 2025-02-15 at 11 55 50 AM" width="1512" src="https://private-user-images.githubusercontent.com/25003283/413570141-d26921c2-054b-4d89-8b21-482aea245c81.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk2ODA5NjMsIm5iZiI6MTczOTY4MDY2MywicGF0aCI6Ii8yNTAwMzI4My80MTM1NzAxNDEtZDI2OTIxYzItMDU0Yi00ZDg5LThiMjEtNDgyYWVhMjQ1YzgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE2VDA0Mzc0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWViNDRjMzg3NmJkOTE1YzgzMjkwNjNkMGU2NDhjYTQ4YWFiNjZmZjExZDU3YmZjNzYwNjU0NTA5NzhkZmViNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e0vnoH2E-RbomJHUG-1jKbXTuUgld0AGFzO8j-W5qHg">

these is happening with ffmpeg, not able to get frames if the video is not finished

> i think the suggestion query is kinda awkward for now, maybe we could tune down the scope of this PR to just simple full text search over OCR/transcriptions or iterate on this UX further as you prefer

yeah I thought this too ,creating suggestions for keywords is taking too long and sometimes the suggestion query doesn't any records.

shaik-zeeshan commented:
https://github.com/user-attachments/assets/8fa1d4bd-d355-4265-a631-002174ec01c4

removed query suggestion, added app based filtering, improve ux while searching, add keyword highlight (not sure if we should keep it)

louis030195 commented:
nice!

i guess just remove the yellow highlight feature, not great 


regarding links i agree it would be useful, you mean in the OCR? is there a way to extract links without AI? eg regex

i think it would be cool if in the backend we store the URL the user is on while on the browser but we don't have this yet

shaik-zeeshan commented:
> i guess just remove the yellow highlight feature, not great

okay

> regarding links i agree it would be useful, you mean in the OCR? is there a way to extract links without AI? eg regex

some browser don't show complete url (like arc, safari)

> i think it would be cool if in the backend we store the URL the user is on while on the browser but we don't have this yet

waiting for this pr #1381

louis030195 commented:
> > i guess just remove the yellow highlight feature, not great
> 
> okay
> 
> > regarding links i agree it would be useful, you mean in the OCR? is there a way to extract links without AI? eg regex
> 
> some browser don't show complete url (like arc, safari)
> 
> > i think it would be cool if in the backend we store the URL the user is on while on the browser but we don't have this yet
> 
> waiting for this pr #1381

i'm pushing soon code to get the browser url data in api result if useful

louis030195 commented:
done

louis030195 commented:
https://cap.so/s/djh6h7dwmtddw5m

for some reason 

```

2025-02-20T16:30:19.143234Z ERROR screenpipe_server::server: Failed to extract frame 92996: ffmpeg process failed: ffmpeg version 7.1 Copyright (c) 2000-2024 the FFmpeg developers
  built with Apple clang version 16.0.0 (clang-1600.0.26.4)
  configuration: --prefix=/opt/homebrew/Cellar/ffmpeg/7.1_4 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay
--enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-li
brav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-
libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype -
-enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enable-libsoxr --enable-libzmq --enable-libzimg --disable-lib
jack --disable-indev=jack --enable-videotoolbox --enable-audiotoolbox --enable-neon
  libavutil      59. 39.100 / 59. 39.100
  libavcodec     61. 19.100 / 61. 19.100
  libavformat    61.  7.100 / 61.  7.100
  libavdevice    61.  3.100 / 61.  3.100
  libavfilter    10.  4.100 / 10.  4.100
  libswscale      8.  3.100 /  8.  3.100
  libswresample   5.  3.100 /  5.  3.100
  libpostproc    58.  3.100 / 58.  3.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/Users/louisbeaumont/.screenpipe/data/monitor_1_2025-02-10_15-56-21.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf61.7.100
  Duration: 00:01:00.00, start: 0.000000, bitrate: 848 kb/s
  Stream #0:0[0x1](und): Video: hevc (Main) (hvc1 / 0x31637668), yuv420p(tv, progressive), 3024x1964, 848 kb/s, 0.50 fps, 0.50 tbr, 16384 tbn (default)
      Metadata:
        handler_name    : VideoHandler
        vendor_id       : [0][0][0][0]
        encoder         : Lavc61.19.100 libx265
Stream mapping:
  Stream #0:0 -> #0:0 (hevc (native) -> mjpeg (native))
Press [q] to stop, [?] for help
[vf#0:0 @ 0x124e045f0] No filtered frames for output stream, trying to initialize anyway.
[mjpeg @ 0x136506ff0] Non full-range YUV is non-standard, set strict_std_compliance to at most unofficial to use it.
[mjpeg @ 0x134f056f0] ff_frame_thread_encoder_init failed
[vost#0:0/mjpeg @ 0x134f05420] Error while opening encoder - maybe incorrect parameters such as bit_rate, rate, width or height.
[vf#0:0 @ 0x600001d00140] Task finished with error code: -22 (Invalid argument)
[vf#0:0 @ 0x600001d00140] Terminating thread with return code -22 (Invalid argument)
[vost#0:0/mjpeg @ 0x134f05420] Could not open encoder before EOF
[vost#0:0/mjpeg @ 0x134f05420] Task finished with error code: -22 (Invalid argument)
[vost#0:0/mjpeg @ 0x134f05420] Terminating thread with return code -22 (Invalid argument)
[out#0/image2 @ 0x600001910000] Nothing was written into output file, because at least one of its streams received no packets.
```

louis030195 commented:
did you base your branch on this? https://github.com/mediar-ai/screenpipe/pull/1364

shaik-zeeshan commented:
> https://cap.so/s/djh6h7dwmtddw5m

solving the issues mentioned in video

the ffmpeg error occures when the video is being record or when video is not saved properly when server is closed

louis030195 commented:
<img width="1512" alt="Screenshot 2025-02-20 at 12 47 11 PM" src="https://github.com/user-attachments/assets/593e4129-64d3-400a-9271-890466347189" />

why are frames failing?

louis030195 commented:
these are old ones not video being wrriten - also ideally should show some skeleton or thing to show its loading 

i guess for recent frames we should prob leverage the streaming api but its not ready yet too experiemntal

louis030195 commented:
<img width="1512" alt="Screenshot 2025-02-20 at 12 48 59 PM" src="https://github.com/user-attachments/assets/a015a3c7-f8bf-4269-aafe-6bed34c27959" />

shaik-zeeshan commented:
fixed ffmpeg error

> these are old ones not video being wrriten - also ideally should show some skeleton or thing to show its loading

sure , will add this

shaik-zeeshan commented:
added loading spinner to images


https://github.com/user-attachments/assets/a0d2ac6e-e73d-4e22-99ba-5c1082bdde4b

louis030195 commented:
/approve

louis030195 commented:
/tip $100 @shaik-zeeshan 

amazing!

neo773 commented:
> added loading spinner to images
> 
>  Screen.Recording.2025-02-21.at.11.10.18.AM.1.mp4

That layout shift at the bottom is kinda annoying, I guess we could render a full row of skeleton images if user is typing?

shaik-zeeshan commented:
> That layout shift at the bottom is kinda annoying, I guess we could render a full row of skeleton images if user is typing?

Sure. We could do this while data is loading.
Will create another pe for this
